termlink/snomedct.py

```diff
@@ -0,0 +1,143 @@
+"""Handles Snomed-CT conversion.
+
+This module provides methods to extract, transform and load relationships
+defined by the snomed-ct dataset.
+
+The download files for snomed-ct are provided at https://www.nlm.nih.gov/healthit/snomedct/.
+"""
+import petl as etl
+import json
+import os
+
+from termlink.commands import SubCommand
+from termlink.models import Coding, Relationship, RelationshipSchema
+from termlink.services import RelationshipService
+from os import listdir
+from dataclasses import replace
+from urllib.parse import urlparse
+
+_STATED_RELATIONSHIP_FILE_FIELDS = ["id", "effectiveTime", "active", "moduleId", "sourceId", "destinationId",
+                                    "relationshipGroup", "typeId", "characteristicTypeId", "modifierId"]
+_DESCRIPTION_FILE_FIELDS = ["id", "effectiveTime", "active", "moduleId", "conceptId", "languageCode", "typeId",
+                            "term", "caseSignificanceId"]
+_IS_A_TYPE_ID = "116680003"
+_FULLY_SPECIFIED_NAME_TYPE_ID = "900000000000003001"
+_FIRST_LETTER_CAPITAL_CASE_ID = "900000000000020002"
+
+
+def _to_json(rec):
+    """
+    Convert record in table to Relationship as a JSON object
+
+    Record is expected to have the following fields: [ source.CODE, source.STR, target.CODE, target.STR]
+
+    Args:
+        rec: A table record
+
+    Returns:
+        A new record containing a single field, which is the JSON object
+    """
+
+    source = Coding(
+        system="http://snomed.info/sct",
+        code=rec['sourceId'],
+        display=rec['source.term']
+    )
+
+    target = Coding(
+        system="http://snomed.info/sct",
+        code=rec['destinationId'],
+        display=rec['target.term']
+    )
+
+    if rec['include-version-mapping']:
+        source = replace(source, version=rec['effectiveTime'])
+        target = replace(target, version=rec['effectiveTime'])
+
+    relationship = Relationship('subsumes', source, target)
+
+    schema = RelationshipSchema()
+    return [json.dumps(schema.dump(relationship))]
+
+
+class Command(SubCommand):
+    """
+    A command executor for Snomed-CT operations
+    """
+
+    @staticmethod
+    def execute(args):
+        """
+        Prints a JSON array of `Relationship` objects to stdout
+
+        Args:
+            args: `argparse` parsed arguments
+        """
+        uri = urlparse(args.uri)
+        service = Service(uri, args.include_version_mapping, args.active_only)
+        table = service.get_relationships()
+        etl.io.totext(table, encoding='utf8', template='{relationship}\n')
+
+
+
+class Service(RelationshipService):
+    """Converts the Snomed-CT database"""
```

Comment:
Proper name is "SNOMED CT". I think it's an acronym.

Reply:
OK